e2e/aws/redshift_test.go

```diff
@@ -55,3 +65,221 @@ func testRedshiftServerless(t *testing.T) {
+
+	// create a new schema with tables that can only be accessed if the
+	// auto roles are granted by Teleport automatically.
+	testSchema := "test_" + randASCII(t, 4)
+	_, err = conn.Exec(ctx, fmt.Sprintf("CREATE SCHEMA %q", testSchema))
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		// users/roles can only be dropped after we drop the schema+table.
+		// So, rather than juggling the order of drops, just attempt to drop
+		// everything as part of test cleanup, regardless of what the test
+		// actually created successfully.
+		_, _ = conn.Exec(ctx, fmt.Sprintf("DROP SCHEMA %q CASCADE", testSchema))
+		_, _ = conn.Exec(ctx, fmt.Sprintf("DROP ROLE %q", autoRole1))
+		_, _ = conn.Exec(ctx, fmt.Sprintf("DROP ROLE %q", autoRole2))
+		_, _ = conn.Exec(ctx, fmt.Sprintf("DROP USER %q", autoUserKeep))
+		_, _ = conn.Exec(ctx, fmt.Sprintf("DROP USER %q", autoUserDrop))
+	})
+	testTable := "ctf" // capture the flag :)
+	_, err = conn.Exec(ctx, fmt.Sprintf("CREATE TABLE %q.%q (id int)", testSchema, testTable))
+	require.NoError(t, err)
+
+	// create the roles that Teleport will auto assign.
+	// role 1 only allows usage of the test schema.
+	// role 2 only allows select of the test table in the test schema.
+	// a user needs to have both roles to select from the test table.
+	_, err = conn.Exec(ctx, fmt.Sprintf("CREATE ROLE %q", autoRole1))
+	require.NoError(t, err)
+	_, err = conn.Exec(ctx, fmt.Sprintf("CREATE ROLE %q", autoRole2))
+	require.NoError(t, err)
+	_, err = conn.Exec(ctx, fmt.Sprintf("GRANT USAGE ON SCHEMA %q TO ROLE %q", testSchema, autoRole1))
+	require.NoError(t, err)
+	_, err = conn.Exec(ctx, fmt.Sprintf("GRANT SELECT ON %q.%q TO ROLE %q", testSchema, testTable, autoRole2))
+	require.NoError(t, err)
+	autoRolesQuery := fmt.Sprintf("select 1 from %q.%q", testSchema, testTable)
+
+	var pgxConnMu sync.Mutex
+	for name, test := range map[string]struct {
+		user            string
+		dbUser          string
+		query           string
+		afterConnTestFn func(t *testing.T)
+	}{
+		"iam role": {
+			user: hostUser,
+			// role/<name> is the syntax we require for redshift IAM role auth
+			dbUser: "role/" + mustGetEnv(t, redshiftIAMDBUserEnv),
+			query:  "select 1",
+		},
+		"existing user": {
+			user:   hostUser,
+			dbUser: adminUser.Name,
+			query:  "select 1",
+		},
+		"auto user keep": {
+			user:   autoUserKeep,
+			dbUser: autoUserKeep,
+			query:  autoRolesQuery,
+			afterConnTestFn: func(t *testing.T) {
+				pgxConnMu.Lock()
+				defer pgxConnMu.Unlock()
+				waitForRedshiftAutoUserDeactivate(t, ctx, conn, autoUserKeep)
+			},
+		},
+		"auto user drop": {
+			user:   autoUserDrop,
+			dbUser: autoUserDrop,
+			query:  autoRolesQuery,
+			afterConnTestFn: func(t *testing.T) {
+				pgxConnMu.Lock()
+				defer pgxConnMu.Unlock()
+				waitForRedshiftAutoUserDrop(t, ctx, conn, autoUserDrop)
+			},
+		},
+	} {
+		test := test
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+			t.Run("connect", func(t *testing.T) {
```

Comment:
I think you could merge those 2 t.Runs into one. Also see the comment below about parallels.

Reply:
see above